<!DOCTYPE html><html><head><meta name="viewport" content="width=device-width"/><meta charSet="utf-8"/><title>JavaScript Error Handling</title><link rel="icon" href="/images/favicon.ico"/><meta name="next-head-count" content="4"/><link rel="preload" href="/_next/static/css/b65b5b41f2379875.css" as="style"/><link rel="stylesheet" href="/_next/static/css/b65b5b41f2379875.css" data-n-g=""/><link rel="preload" href="/_next/static/css/0d9654b911e08999.css" as="style"/><link rel="stylesheet" href="/_next/static/css/0d9654b911e08999.css" data-n-p=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-5cd94c89d3acac5f.js"></script><script src="/_next/static/chunks/webpack-69bfa6990bb9e155.js" defer=""></script><script src="/_next/static/chunks/framework-fc97f3f1282ce3ed.js" defer=""></script><script src="/_next/static/chunks/main-f4ae3437c92c1efc.js" defer=""></script><script src="/_next/static/chunks/pages/_app-85d7488a393e293e.js" defer=""></script><script src="/_next/static/chunks/211-ca3cd870e26e881a.js" defer=""></script><script src="/_next/static/chunks/pages/notes/%5B...noteKey%5D-bc6552c3ae1e3718.js" defer=""></script><script src="/_next/static/TLVs-fHNJx4QzHIHhPfMV/_buildManifest.js" defer=""></script><script src="/_next/static/TLVs-fHNJx4QzHIHhPfMV/_ssgManifest.js" defer=""></script><script src="/_next/static/TLVs-fHNJx4QzHIHhPfMV/_middlewareManifest.js" defer=""></script></head><body><div id="__next"><section class="Layout_centeredSection__nmU9U"><ul class="NavigationBar_navbar__CBMoV"><li class="NavigationBar_navitem__LKB5F"><a class="NavigationBar_navitemContent__jEyWq" href="/">Home</a><span class="NavigationBar_separator__qvEVD">|</span></li><li class="NavigationBar_navitem__LKB5F"><button class="NavigationBar_navitemContent__jEyWq">Back</button><span class="NavigationBar_separator__qvEVD">|</span></li></ul><h1 class="NoteFormattedContent_noteTitle__Oi9sD">JavaScript Error Handling</h1><article class="NoteFormattedContent_note__8cHeE">
<ul>
  <li><a href="#error-object">Error Object</a></li>
  <li><a href="#error-handling">Error Handling</a></li>
  <li><a href="#async-error-handling">Async Error Handling</a></li>
  <li><a href="#error-event">Error Event</a></li>
  <li><a href="#custom-errors">Custom Errors</a></li>
  <li><a href="#references">References</a></li>
</ul>
<p>
  When on some part of code execution unexpected situation happens will be created and thrown <code>Error</code> object. Then going up by call stack <code>JavaScript</code> interpreter will check for <code>try/catch</code> logic on each next
  level until the end of call stack resulting to termination of the program.
</p>
<h2>Error Object</h2>
<p><code>Error</code> is a most basic type of exception object in <code>JavaScript</code>. Also this object is often used as a custom error objects prototype.</p>
<p>Standard forces exception to have at least two fields:</p>
<ul>
  <li><code>name</code> - defined type of error</li>
  <li><code>message</code> - human readable description of exception, which should give some error context</li>
</ul>
<p>On most modern browsers <code>Error</code> object contains <code>stack</code> field, which contains information about call stack of execution, which may be useful to get a context of exception.</p>
<p><strong>Built-in <code>JS</code> error types:</strong></p>
<ul>
  <li>
    <p><code>RangeError</code> - usually thrown, when out of expected input provided. e.g. following code will throw <code>RangeError</code>:</p>
    <pre><code class="hljs language-js"><span class="hljs-keyword">var</span> pi = <span class="hljs-number">3.14159</span>
pi.<span class="hljs-title function_">toFixed</span>(<span class="hljs-number">100000</span>)
</code></pre>
  </li>
  <li>
    <p><code>ReferenceError</code> - is thrown on attempt to access non existing variable</p>
  </li>
  <li>
    <p><code>SyntaxError</code> - is thrown, when syntactically invalid provided to the <code>JS</code> interpreter</p>
  </li>
  <li>
    <p><code>TypeError</code> - thrown, when there is an attempt to use a type in a not appropriate way. e.g. trying to execute non-existing method</p>
  </li>
  <li>
    <p><code>URIError</code> - thrown by <code>encodeURI</code> or <code>decodeURI</code>, when not valid <code>url</code> provided</p>
  </li>
  <li>
    <p><code>EvalError</code> - thrown by <code>eval</code> function</p>
  </li>
</ul>
<h2>Error Handling</h2>
<p>There is a special construct to handle errors in <code>JavaScript</code>:</p>
<pre><code class="hljs language-js"><span class="hljs-keyword">try</span> {
  <span class="hljs-comment">// executing code</span>
} <span class="hljs-keyword">catch</span> (err) {
  <span class="hljs-comment">// handling possible error</span>
} <span class="hljs-keyword">finally</span> {
  <span class="hljs-comment">// this will be executed both in success and error cases</span>
}
</code></pre>
<p><em>catch</em> block stops propagation of error further up on the call stack and allows to react to the error happened.</p>
<p>When there is a need to differentiate reaction on different kinds of errors it is possible to use <code>instanceof</code> operator to distinguish error type:</p>
<pre><code class="hljs language-js"><span class="hljs-keyword">try</span> {
  <span class="hljs-comment">// ...</span>
} <span class="hljs-keyword">catch</span> (err) {
  <span class="hljs-keyword">if</span> (err <span class="hljs-keyword">instanceof</span> <span class="hljs-title class_">TypeError</span>) {
    <span class="hljs-comment">// handling TypeError</span>
  } <span class="hljs-keyword">else</span> <span class="hljs-keyword">if</span> (err <span class="hljs-keyword">instanceof</span> <span class="hljs-title class_">ReferenceError</span>) {
    <span class="hljs-comment">// handling ReferenceError</span>
  } <span class="hljs-keyword">else</span> {
    <span class="hljs-comment">// handle any other error type, e.g.</span>
    <span class="hljs-keyword">throw</span> err
  }
}
</code></pre>
<p>It is possible to re-throw caught error any number of times.</p>
<h2>Async Error Handling</h2>
<p>When dealing with asynchronous code <code>try/catch</code> clause may not catch possible exceptions. This may happen because asynchronous function running outside of call stack, calling it.</p>
<pre><code class="hljs language-js"><span class="hljs-keyword">try</span> {
  <span class="hljs-title function_">throwingAsyncFunc</span>() <span class="hljs-comment">// this function throws exception asynchronously</span>
} <span class="hljs-keyword">catch</span> (err) {
  <span class="hljs-comment">// exception will not be caught here</span>
}
</code></pre>
<p>If <code>throwingAsyncFunc</code> returns <code>Promise</code> it is possible to use <code>async/await</code> pattern to handle it's errors with <code>try/catch</code> block:</p>
<pre><code class="hljs language-js"><span class="hljs-keyword">async</span> <span class="hljs-keyword">function</span> <span class="hljs-title function_">handler</span>(<span class="hljs-params"></span>) {
  <span class="hljs-keyword">try</span> {
    <span class="hljs-keyword">await</span> <span class="hljs-title function_">throwingAsyncFunc</span>() <span class="hljs-comment">// this function throws exception asynchronously</span>
  } <span class="hljs-keyword">catch</span> (err) {
    <span class="hljs-comment">// exception will be caught here</span>
  }
}
</code></pre>
<h2>Error Event</h2>
<p>In <code>JavaScript</code> errors are also events. So it is possible to react on this events. For example in browser api it is possible to do with following event listener:</p>
<pre><code class="hljs language-js"><span class="hljs-variable language_">window</span>.<span class="hljs-title function_">addEventListener</span>(<span class="hljs-string">"error"</span>, <span class="hljs-keyword">function</span> (<span class="hljs-params">e</span>) {
  <span class="hljs-variable language_">console</span>.<span class="hljs-title function_">log</span>(e.<span class="hljs-property">error</span>)
})
</code></pre>
<p>
  Such error handler will catch unhandled errors from any execution context.
  Moreover, <code>addEventListener</code> allows to attach to the <code>error</code> event variety of different handlers.
  This allows to create many centralized handlers for different error types.
</p>
<p>For <code>Node.Js</code>:</p>
<pre><code class="hljs language-js">process.<span class="hljs-title function_">on</span>(<span class="hljs-string">"uncaughtException"</span>, <span class="hljs-function"><span class="hljs-params">error</span> =></span> {
  <span class="hljs-comment">// handle error</span>
})
</code></pre>
<p>It is also useful to use error events and not directly throw an error when dealing with streams.</p>
<pre><code class="hljs language-js">source.<span class="hljs-title function_">run</span>(<span class="hljs-function"><span class="hljs-params">stream</span> =></span> {
  stream
    .<span class="hljs-title function_">on</span>(<span class="hljs-string">"data"</span>, <span class="hljs-function"><span class="hljs-params">data</span> =></span> {
      <span class="hljs-comment">// handle arriving piece of data</span>
    })
    .<span class="hljs-title function_">on</span>(<span class="hljs-string">"end"</span>, <span class="hljs-function"><span class="hljs-params">result</span> =></span> {
      <span class="hljs-comment">// handling end of stream</span>
    })
    .<span class="hljs-title function_">on</span>(<span class="hljs-string">"error"</span>, <span class="hljs-function"><span class="hljs-params">err</span> =></span> {
      <span class="hljs-comment">// handling possible error during the stream</span>
    })
})
</code></pre>
<h2>Custom Errors</h2>
<p>If there is a need to catch and re-throw some custom exception up it is a good practice to maintain a context of previous exception:</p>
<pre><code class="hljs language-js"><span class="hljs-keyword">class</span> <span class="hljs-title class_">CustomError</span> <span class="hljs-keyword">extends</span> <span class="hljs-title class_ inherited__">Error</span> {
  <span class="hljs-title function_">constructor</span>(<span class="hljs-params">message, err</span>) {
    <span class="hljs-variable language_">super</span>(message)

    <span class="hljs-keyword">if</span> (<span class="hljs-title class_">Error</span>.<span class="hljs-property">captureStackTrace</span>) {
      <span class="hljs-title class_">Error</span>.<span class="hljs-title function_">captureStackTrace</span>(<span class="hljs-variable language_">this</span>, <span class="hljs-title class_">CustomError</span>)
    }
    <span class="hljs-variable language_">this</span>.<span class="hljs-property">wrappedError</span> = err
    <span class="hljs-variable language_">this</span>.<span class="hljs-property">name</span> = <span class="hljs-string">"CustomError"</span>
    <span class="hljs-variable language_">this</span>.<span class="hljs-property">date</span> = <span class="hljs-keyword">new</span> <span class="hljs-title class_">Date</span>()
  }
}
</code></pre>
<h2>References</h2>
<ul>
  <li><a href="https://blog.bitsrc.io/errors-handling-for-vue-class-components-2f152f7c7515">Better errors handling for ES/Typescript classes</a></li>
  <li><a href="https://www.sitepoint.com/proper-error-handling-javascript/">A Guide to Proper Error Handling in JavaScript</a></li>
  <li><a href="https://sematext.com/blog/node-js-error-handling/#toc-types-of-errors-operational-vs-programmer-errors-2">Node.js Error Handling Made Easy: Best Practices On Just About Everything You Need to Know</a></li>
  <li><a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error">MDN Error</a></li>
</ul>
</article></section></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"noteKey":["javascript","js-error-handling"],"note":{"title":"JavaScript Error Handling","date":"2022-04-13","content":"\n\u003cul\u003e\n  \u003cli\u003e\u003ca href=\"#error-object\"\u003eError Object\u003c/a\u003e\u003c/li\u003e\n  \u003cli\u003e\u003ca href=\"#error-handling\"\u003eError Handling\u003c/a\u003e\u003c/li\u003e\n  \u003cli\u003e\u003ca href=\"#async-error-handling\"\u003eAsync Error Handling\u003c/a\u003e\u003c/li\u003e\n  \u003cli\u003e\u003ca href=\"#error-event\"\u003eError Event\u003c/a\u003e\u003c/li\u003e\n  \u003cli\u003e\u003ca href=\"#custom-errors\"\u003eCustom Errors\u003c/a\u003e\u003c/li\u003e\n  \u003cli\u003e\u003ca href=\"#references\"\u003eReferences\u003c/a\u003e\u003c/li\u003e\n\u003c/ul\u003e\n\u003cp\u003e\n  When on some part of code execution unexpected situation happens will be created and thrown \u003ccode\u003eError\u003c/code\u003e object. Then going up by call stack \u003ccode\u003eJavaScript\u003c/code\u003e interpreter will check for \u003ccode\u003etry/catch\u003c/code\u003e logic on each next\n  level until the end of call stack resulting to termination of the program.\n\u003c/p\u003e\n\u003ch2\u003eError Object\u003c/h2\u003e\n\u003cp\u003e\u003ccode\u003eError\u003c/code\u003e is a most basic type of exception object in \u003ccode\u003eJavaScript\u003c/code\u003e. Also this object is often used as a custom error objects prototype.\u003c/p\u003e\n\u003cp\u003eStandard forces exception to have at least two fields:\u003c/p\u003e\n\u003cul\u003e\n  \u003cli\u003e\u003ccode\u003ename\u003c/code\u003e - defined type of error\u003c/li\u003e\n  \u003cli\u003e\u003ccode\u003emessage\u003c/code\u003e - human readable description of exception, which should give some error context\u003c/li\u003e\n\u003c/ul\u003e\n\u003cp\u003eOn most modern browsers \u003ccode\u003eError\u003c/code\u003e object contains \u003ccode\u003estack\u003c/code\u003e field, which contains information about call stack of execution, which may be useful to get a context of exception.\u003c/p\u003e\n\u003cp\u003e\u003cstrong\u003eBuilt-in \u003ccode\u003eJS\u003c/code\u003e error types:\u003c/strong\u003e\u003c/p\u003e\n\u003cul\u003e\n  \u003cli\u003e\n    \u003cp\u003e\u003ccode\u003eRangeError\u003c/code\u003e - usually thrown, when out of expected input provided. e.g. following code will throw \u003ccode\u003eRangeError\u003c/code\u003e:\u003c/p\u003e\n    \u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-keyword\"\u003evar\u003c/span\u003e pi = \u003cspan class=\"hljs-number\"\u003e3.14159\u003c/span\u003e\npi.\u003cspan class=\"hljs-title function_\"\u003etoFixed\u003c/span\u003e(\u003cspan class=\"hljs-number\"\u003e100000\u003c/span\u003e)\n\u003c/code\u003e\u003c/pre\u003e\n  \u003c/li\u003e\n  \u003cli\u003e\n    \u003cp\u003e\u003ccode\u003eReferenceError\u003c/code\u003e - is thrown on attempt to access non existing variable\u003c/p\u003e\n  \u003c/li\u003e\n  \u003cli\u003e\n    \u003cp\u003e\u003ccode\u003eSyntaxError\u003c/code\u003e - is thrown, when syntactically invalid provided to the \u003ccode\u003eJS\u003c/code\u003e interpreter\u003c/p\u003e\n  \u003c/li\u003e\n  \u003cli\u003e\n    \u003cp\u003e\u003ccode\u003eTypeError\u003c/code\u003e - thrown, when there is an attempt to use a type in a not appropriate way. e.g. trying to execute non-existing method\u003c/p\u003e\n  \u003c/li\u003e\n  \u003cli\u003e\n    \u003cp\u003e\u003ccode\u003eURIError\u003c/code\u003e - thrown by \u003ccode\u003eencodeURI\u003c/code\u003e or \u003ccode\u003edecodeURI\u003c/code\u003e, when not valid \u003ccode\u003eurl\u003c/code\u003e provided\u003c/p\u003e\n  \u003c/li\u003e\n  \u003cli\u003e\n    \u003cp\u003e\u003ccode\u003eEvalError\u003c/code\u003e - thrown by \u003ccode\u003eeval\u003c/code\u003e function\u003c/p\u003e\n  \u003c/li\u003e\n\u003c/ul\u003e\n\u003ch2\u003eError Handling\u003c/h2\u003e\n\u003cp\u003eThere is a special construct to handle errors in \u003ccode\u003eJavaScript\u003c/code\u003e:\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e {\n  \u003cspan class=\"hljs-comment\"\u003e// executing code\u003c/span\u003e\n} \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e (err) {\n  \u003cspan class=\"hljs-comment\"\u003e// handling possible error\u003c/span\u003e\n} \u003cspan class=\"hljs-keyword\"\u003efinally\u003c/span\u003e {\n  \u003cspan class=\"hljs-comment\"\u003e// this will be executed both in success and error cases\u003c/span\u003e\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e\u003cem\u003ecatch\u003c/em\u003e block stops propagation of error further up on the call stack and allows to react to the error happened.\u003c/p\u003e\n\u003cp\u003eWhen there is a need to differentiate reaction on different kinds of errors it is possible to use \u003ccode\u003einstanceof\u003c/code\u003e operator to distinguish error type:\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e {\n  \u003cspan class=\"hljs-comment\"\u003e// ...\u003c/span\u003e\n} \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e (err) {\n  \u003cspan class=\"hljs-keyword\"\u003eif\u003c/span\u003e (err \u003cspan class=\"hljs-keyword\"\u003einstanceof\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eTypeError\u003c/span\u003e) {\n    \u003cspan class=\"hljs-comment\"\u003e// handling TypeError\u003c/span\u003e\n  } \u003cspan class=\"hljs-keyword\"\u003eelse\u003c/span\u003e \u003cspan class=\"hljs-keyword\"\u003eif\u003c/span\u003e (err \u003cspan class=\"hljs-keyword\"\u003einstanceof\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eReferenceError\u003c/span\u003e) {\n    \u003cspan class=\"hljs-comment\"\u003e// handling ReferenceError\u003c/span\u003e\n  } \u003cspan class=\"hljs-keyword\"\u003eelse\u003c/span\u003e {\n    \u003cspan class=\"hljs-comment\"\u003e// handle any other error type, e.g.\u003c/span\u003e\n    \u003cspan class=\"hljs-keyword\"\u003ethrow\u003c/span\u003e err\n  }\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003eIt is possible to re-throw caught error any number of times.\u003c/p\u003e\n\u003ch2\u003eAsync Error Handling\u003c/h2\u003e\n\u003cp\u003eWhen dealing with asynchronous code \u003ccode\u003etry/catch\u003c/code\u003e clause may not catch possible exceptions. This may happen because asynchronous function running outside of call stack, calling it.\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e {\n  \u003cspan class=\"hljs-title function_\"\u003ethrowingAsyncFunc\u003c/span\u003e() \u003cspan class=\"hljs-comment\"\u003e// this function throws exception asynchronously\u003c/span\u003e\n} \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e (err) {\n  \u003cspan class=\"hljs-comment\"\u003e// exception will not be caught here\u003c/span\u003e\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003eIf \u003ccode\u003ethrowingAsyncFunc\u003c/code\u003e returns \u003ccode\u003ePromise\u003c/code\u003e it is possible to use \u003ccode\u003easync/await\u003c/code\u003e pattern to handle it's errors with \u003ccode\u003etry/catch\u003c/code\u003e block:\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-keyword\"\u003easync\u003c/span\u003e \u003cspan class=\"hljs-keyword\"\u003efunction\u003c/span\u003e \u003cspan class=\"hljs-title function_\"\u003ehandler\u003c/span\u003e(\u003cspan class=\"hljs-params\"\u003e\u003c/span\u003e) {\n  \u003cspan class=\"hljs-keyword\"\u003etry\u003c/span\u003e {\n    \u003cspan class=\"hljs-keyword\"\u003eawait\u003c/span\u003e \u003cspan class=\"hljs-title function_\"\u003ethrowingAsyncFunc\u003c/span\u003e() \u003cspan class=\"hljs-comment\"\u003e// this function throws exception asynchronously\u003c/span\u003e\n  } \u003cspan class=\"hljs-keyword\"\u003ecatch\u003c/span\u003e (err) {\n    \u003cspan class=\"hljs-comment\"\u003e// exception will be caught here\u003c/span\u003e\n  }\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003ch2\u003eError Event\u003c/h2\u003e\n\u003cp\u003eIn \u003ccode\u003eJavaScript\u003c/code\u003e errors are also events. So it is possible to react on this events. For example in browser api it is possible to do with following event listener:\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-variable language_\"\u003ewindow\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003eaddEventListener\u003c/span\u003e(\u003cspan class=\"hljs-string\"\u003e\"error\"\u003c/span\u003e, \u003cspan class=\"hljs-keyword\"\u003efunction\u003c/span\u003e (\u003cspan class=\"hljs-params\"\u003ee\u003c/span\u003e) {\n  \u003cspan class=\"hljs-variable language_\"\u003econsole\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003elog\u003c/span\u003e(e.\u003cspan class=\"hljs-property\"\u003eerror\u003c/span\u003e)\n})\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e\n  Such error handler will catch unhandled errors from any execution context.\n  Moreover, \u003ccode\u003eaddEventListener\u003c/code\u003e allows to attach to the \u003ccode\u003eerror\u003c/code\u003e event variety of different handlers.\n  This allows to create many centralized handlers for different error types.\n\u003c/p\u003e\n\u003cp\u003eFor \u003ccode\u003eNode.Js\u003c/code\u003e:\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003eprocess.\u003cspan class=\"hljs-title function_\"\u003eon\u003c/span\u003e(\u003cspan class=\"hljs-string\"\u003e\"uncaughtException\"\u003c/span\u003e, \u003cspan class=\"hljs-function\"\u003e\u003cspan class=\"hljs-params\"\u003eerror\u003c/span\u003e =\u003e\u003c/span\u003e {\n  \u003cspan class=\"hljs-comment\"\u003e// handle error\u003c/span\u003e\n})\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003eIt is also useful to use error events and not directly throw an error when dealing with streams.\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003esource.\u003cspan class=\"hljs-title function_\"\u003erun\u003c/span\u003e(\u003cspan class=\"hljs-function\"\u003e\u003cspan class=\"hljs-params\"\u003estream\u003c/span\u003e =\u003e\u003c/span\u003e {\n  stream\n    .\u003cspan class=\"hljs-title function_\"\u003eon\u003c/span\u003e(\u003cspan class=\"hljs-string\"\u003e\"data\"\u003c/span\u003e, \u003cspan class=\"hljs-function\"\u003e\u003cspan class=\"hljs-params\"\u003edata\u003c/span\u003e =\u003e\u003c/span\u003e {\n      \u003cspan class=\"hljs-comment\"\u003e// handle arriving piece of data\u003c/span\u003e\n    })\n    .\u003cspan class=\"hljs-title function_\"\u003eon\u003c/span\u003e(\u003cspan class=\"hljs-string\"\u003e\"end\"\u003c/span\u003e, \u003cspan class=\"hljs-function\"\u003e\u003cspan class=\"hljs-params\"\u003eresult\u003c/span\u003e =\u003e\u003c/span\u003e {\n      \u003cspan class=\"hljs-comment\"\u003e// handling end of stream\u003c/span\u003e\n    })\n    .\u003cspan class=\"hljs-title function_\"\u003eon\u003c/span\u003e(\u003cspan class=\"hljs-string\"\u003e\"error\"\u003c/span\u003e, \u003cspan class=\"hljs-function\"\u003e\u003cspan class=\"hljs-params\"\u003eerr\u003c/span\u003e =\u003e\u003c/span\u003e {\n      \u003cspan class=\"hljs-comment\"\u003e// handling possible error during the stream\u003c/span\u003e\n    })\n})\n\u003c/code\u003e\u003c/pre\u003e\n\u003ch2\u003eCustom Errors\u003c/h2\u003e\n\u003cp\u003eIf there is a need to catch and re-throw some custom exception up it is a good practice to maintain a context of previous exception:\u003c/p\u003e\n\u003cpre\u003e\u003ccode class=\"hljs language-js\"\u003e\u003cspan class=\"hljs-keyword\"\u003eclass\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eCustomError\u003c/span\u003e \u003cspan class=\"hljs-keyword\"\u003eextends\u003c/span\u003e \u003cspan class=\"hljs-title class_ inherited__\"\u003eError\u003c/span\u003e {\n  \u003cspan class=\"hljs-title function_\"\u003econstructor\u003c/span\u003e(\u003cspan class=\"hljs-params\"\u003emessage, err\u003c/span\u003e) {\n    \u003cspan class=\"hljs-variable language_\"\u003esuper\u003c/span\u003e(message)\n\n    \u003cspan class=\"hljs-keyword\"\u003eif\u003c/span\u003e (\u003cspan class=\"hljs-title class_\"\u003eError\u003c/span\u003e.\u003cspan class=\"hljs-property\"\u003ecaptureStackTrace\u003c/span\u003e) {\n      \u003cspan class=\"hljs-title class_\"\u003eError\u003c/span\u003e.\u003cspan class=\"hljs-title function_\"\u003ecaptureStackTrace\u003c/span\u003e(\u003cspan class=\"hljs-variable language_\"\u003ethis\u003c/span\u003e, \u003cspan class=\"hljs-title class_\"\u003eCustomError\u003c/span\u003e)\n    }\n    \u003cspan class=\"hljs-variable language_\"\u003ethis\u003c/span\u003e.\u003cspan class=\"hljs-property\"\u003ewrappedError\u003c/span\u003e = err\n    \u003cspan class=\"hljs-variable language_\"\u003ethis\u003c/span\u003e.\u003cspan class=\"hljs-property\"\u003ename\u003c/span\u003e = \u003cspan class=\"hljs-string\"\u003e\"CustomError\"\u003c/span\u003e\n    \u003cspan class=\"hljs-variable language_\"\u003ethis\u003c/span\u003e.\u003cspan class=\"hljs-property\"\u003edate\u003c/span\u003e = \u003cspan class=\"hljs-keyword\"\u003enew\u003c/span\u003e \u003cspan class=\"hljs-title class_\"\u003eDate\u003c/span\u003e()\n  }\n}\n\u003c/code\u003e\u003c/pre\u003e\n\u003ch2\u003eReferences\u003c/h2\u003e\n\u003cul\u003e\n  \u003cli\u003e\u003ca href=\"https://blog.bitsrc.io/errors-handling-for-vue-class-components-2f152f7c7515\"\u003eBetter errors handling for ES/Typescript classes\u003c/a\u003e\u003c/li\u003e\n  \u003cli\u003e\u003ca href=\"https://www.sitepoint.com/proper-error-handling-javascript/\"\u003eA Guide to Proper Error Handling in JavaScript\u003c/a\u003e\u003c/li\u003e\n  \u003cli\u003e\u003ca href=\"https://sematext.com/blog/node-js-error-handling/#toc-types-of-errors-operational-vs-programmer-errors-2\"\u003eNode.js Error Handling Made Easy: Best Practices On Just About Everything You Need to Know\u003c/a\u003e\u003c/li\u003e\n  \u003cli\u003e\u003ca href=\"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error\"\u003eMDN Error\u003c/a\u003e\u003c/li\u003e\n\u003c/ul\u003e\n"}},"__N_SSG":true},"page":"/notes/[...noteKey]","query":{"noteKey":["javascript","js-error-handling"]},"buildId":"TLVs-fHNJx4QzHIHhPfMV","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>